internal/cloud/backend_test.go

```diff
@@ -146,6 +146,124 @@ func TestCloud_PrepareConfig(t *testing.T) {
+				"TF_ORGANIZATION": "example-org",
+			},
+		},
+		"with no organization attribute or env var": {
+			config: cty.ObjectVal(map[string]cty.Value{
+				"organization": cty.NullVal(cty.String),
+				"workspaces": cty.ObjectVal(map[string]cty.Value{
+					"name": cty.StringVal("prod"),
+					"tags": cty.NullVal(cty.Set(cty.String)),
+				}),
+			}),
+			vars:        map[string]string{},
+			expectedErr: `Invalid organization value: The "organization" attribute value must not be empty.`,
+		},
+	}
+
+	for name, tc := range cases {
+		s := testServer(t)
+		b := New(testDisco(s))
+
+		for k, v := range tc.vars {
+			os.Setenv(k, v)
+			defer os.Unsetenv(k)
+		}
+
+		_, valDiags := b.PrepareConfig(tc.config)
+		if valDiags.Err() != nil && tc.expectedErr != "" {
+			actualErr := valDiags.Err().Error()
+			if !strings.Contains(actualErr, tc.expectedErr) {
+				t.Fatalf("%s: unexpected validation result: %v", name, valDiags.Err())
+			}
+		}
+	}
+}
+
+func TestCloud_configWithEnvVars(t *testing.T) {
+	cases := map[string]struct {
+		config cty.Value
+		vars   map[string]string
+	}{
+		"with no organization specified": {
+			config: cty.ObjectVal(map[string]cty.Value{
+				"hostname":     cty.NullVal(cty.String),
+				"token":        cty.NullVal(cty.String),
+				"organization": cty.NullVal(cty.String),
+				"workspaces": cty.ObjectVal(map[string]cty.Value{
+					"name": cty.StringVal("prod"),
+					"tags": cty.NullVal(cty.Set(cty.String)),
+				}),
+			}),
+			vars: map[string]string{
+				"TF_ORGANIZATION": "hashicorp",
+			},
+		},
+		"with both organization and env var specified": {
+			config: cty.ObjectVal(map[string]cty.Value{
+				"hostname":     cty.NullVal(cty.String),
+				"token":        cty.NullVal(cty.String),
+				"organization": cty.StringVal("hashicorp"),
+				"workspaces": cty.ObjectVal(map[string]cty.Value{
+					"name": cty.StringVal("prod"),
+					"tags": cty.NullVal(cty.Set(cty.String)),
+				}),
+			}),
+			vars: map[string]string{
+				"TF_ORGANIZATION": "we-should-not-see-this",
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		s := testServer(t)
+		b := New(testDisco(s))
+
+		for k, v := range tc.vars {
+			os.Setenv(k, v)
+			defer os.Unsetenv(k)
+		}
+
+		_, valDiags := b.PrepareConfig(tc.config)
+		if valDiags.Err() != nil {
+			t.Fatalf("%s: unexpected validation result: %v", name, valDiags.Err())
+		}
+
+		diags := b.Configure(tc.config)
+		if diags.Err() != nil {
+			t.Fatalf("%s: unexpected configuration result: %v", name, valDiags.Err())
+		}
+
+		for k, v := range tc.vars {
+			switch k {
+			case "TF_ORGANIZATION":
+				val := tc.config.GetAttr("organization")
+				if !val.IsNull() && val.AsString() != "" && val.AsString() != b.organization {
+					t.Fatalf("%s: organization not valid: %s, expected: %s", name, b.organization, val.AsString())
+				} else if val.IsNull() && b.organization != v {
+					t.Fatalf("%s: organization not valid: %s, expected: %s", name, b.organization, v)
+				}
+			}
```

Comment:
I found the logic in this test hard to follow, as it's just about as complex as the logic it's trying to cover. I wonder if we could alter the table used to derive the tests in order to make this simpler.

A common pattern in Terraform tests is a `wantFoo` field in the table struct, defining an expected output value. For these tests, maybe a `wantOrganization` could specify what we expect the code to do with the given inputs, so that the test code is just checking that `b.organization == tc.wantOrganization`. Does that make sense to you?